features/network_failures.feature

```diff
@@ -0,0 +1,104 @@
+@uses.config.contract_token
+Feature: Ensure network errors are handled gracefully across various services
+
+  Scenario Outline: Various HTTP errors are handled gracefully on attaching contract token
+    # This test simulates various HTTP errors by mocking the response from the serviceclient
+    # when trying to attach contract token
+    Given a `<release>` `<machine_type>` machine with ubuntu-advantage-tools installed
+    When I create the file `/tmp/response-overlay.json` with the following:
+      """
+      {
+        "https://contracts.canonical.com/v1/context/machines/token": [
+          {
+            "code": <response_code>,
+            "response": {
+              "error": <error_message>
+            }
+          }
+        ]
+      }
+      """
+    And I append the following on uaclient config:
+      """
+      features:
+      serviceclient_url_responses: "/tmp/response-overlay.json"
+      """
+    When I attempt to attach `contract_token` with sudo
+    Then stderr contains substring:
+      """
+      Error connecting to /v1/context/machines/token: <response_code> {"error": <error_message>}
+      """
+    Then the machine is unattached
+
+    Examples: ubuntu release
+      | release | machine_type  | response_code | error_message |
+      | focal   | lxd-container | 400           | "Bad Request" |
+      | focal   | lxd-container | 404           | "Not Found"   |
+      | focal   | lxd-container | 503           | "Bad Gateway" |
+      | jammy   | lxd-container | 400           | "Bad Request" |
+      | jammy   | lxd-container | 404           | "Not Found"   |
+      | jammy   | lxd-container | 503           | "Bad Gateway" |
```

Comment:
Here I was thinking, instead of spinning up a new container for each message, maybe we could just have all the error message overlays created in the test itself. This could make the test just run all the error messages for a single release rather than create a new container per check. 

However, this approach also works and is good by me. Any thoughts? @a-dubs

Reply:
I had that same thought.....

Reply:
We can do that here then, would make it easier to test too :eyes: